File: Go/AutoReply.go
package main

import (
	"bytes"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
)

// Replace these with your own values
const clientID = "your_client_id"
const clientSecret = "your_client_secret"
const tenantID = "your_tenant_id"
const csvFilePath = "auto_reply_messages.csv"

type autoReplyMessage struct {
	UserPrincipalName string
	ExternalMessage   string
	InternalMessage   string
}

// Function to get an access token
func getAccessToken() (string, error) {
	// ... (same as previous example)
}

// Function to set the auto-reply message
func setAutoReplyMessage(accessToken string, message *autoReplyMessage) error {
	url := fmt.Sprintf("https://graph.microsoft.com/v1.0/%s/mailboxSettings/automaticRepliesSetting", message.UserPrincipalName)
	data := map[string]interface{}{
		"status":               "Scheduled",
		"externalReplyMessage": message.ExternalMessage,
		"internalReplyMessage": message.InternalMessage,
	}

	jsonData, err := json.Marshal(data)
	if err != nil {
		return err
	}

	req, err := http.NewRequest("PATCH", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return err
	}
	req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", accessToken))
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{}
	resp, err := client.Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("error setting auto-reply message: %d", resp.StatusCode)
	}

	return nil
}

func main() {
	accessToken, err := getAccessToken()
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error getting access token: %v\n", err)
		return
	}

	file, err := os.Open(csvFilePath)
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error opening CSV file: %v\n", err)
		return
	}
	defer file.Close()

	reader := csv.NewReader(file)
	for {
		row, err := reader.Read()
		if err == io.EOF {
			break
		}
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error reading CSV file: %v\n", err)
			return
		}

		message := &autoReplyMessage{
			UserPrincipalName: row[0],
			ExternalMessage:   row[1],
			InternalMessage:   row[2],
		}

		err = setAutoReplyMessage(accessToken, message)
		if err != nil {
			fmt.Fprintf(os.Stderr, "Error setting auto-reply message for %s: %v\n", message.UserPrincipalName, err)
			continue
		}

		fmt.Printf("Auto-reply message set successfully for %s\n", message.UserPrincipalName)
	}
}

File: Cpp/AutoReply.cpp
#include <iostream>
#include <fstream>
#include <string>
#include <vector>
#include <cpprest/http_client.h>
#include <cpprest/json.h>
#include <cpprest/uri.h>
#include <sstream>

using namespace std;
using namespace web;
using namespace web::http;
using namespace web::http::client;

const string client_id = "your_client_id";
const string client_secret = "your_client_secret";
const string tenant_id = "your_tenant_id";
const string csv_file_path = "auto_reply_messages.csv";

string get_access_token() {
    http_client client(U("https://login.microsoftonline.com/" + tenant_id + "/oauth2/v2.0/token"));
    
    json::value data;
    data[U("grant_type")] = json::value::string(U("client_credentials"));
    data[U("client_id")] = json::value::string(U(client_id));
    data[U("client_secret")] = json::value::string(U(client_secret));
    data[U("scope")] = json::value::string(U("https://graph.microsoft.com/.default"));

    client.request(methods::POST, U(""), data, { { U("Content-Type"), U("application/x-www-form-urlencoded") } })
        .then([](http_response response) {
            if (response.status_code() == status_codes::OK) {
                return response.extract_json();
            }
            throw runtime_error("Failed to get access token.");
        })
        .then([](json::value jsonResponse) {
            return jsonResponse[U("access_token")].as_string();
        })
        .wait();
}

void set_auto_reply_message(const string& access_token, const string& user_principal_name, const string& external_message, const string& internal_message) {
    http_client client(U("https://graph.microsoft.com/v1.0/" + user_principal_name + "/mailboxSettings/automaticRepliesSetting"));
    
    json::value data;
    data[U("status")] = json::value::string(U("Scheduled"));
    data[U("externalReplyMessage")] = json::value::string(U(external_message));
    data[U("internalReplyMessage")] = json::value::string(U(internal_message));

    client.request(methods::PATCH, U(""), data, { { U("Authorization"), U("Bearer " + access_token) }, { U("Content-Type"), U("application/json") } })
        .then([](http_response response) {
            if (response.status_code() != status_codes::OK) {
                throw runtime_error("Failed to set auto-reply message.");
            }
        })
        .wait();
}

void read_csv_and_set_replies() {
    try {
        string access_token = get_access_token();

        ifstream csv_file(csv_file_path);
        string line;
        getline(csv_file, line); // Read header line

        while (getline(csv_file, line)) {
            stringstream ss(line);
            string user_principal_name, external_message, internal_message;

            getline(ss, user_principal_name, ',');
            getline(ss, external_message, ',');
            getline(ss, internal_message, ',');

            set_auto_reply_message(access_token, user_principal_name, external_message, internal_message);
            cout << "Auto-reply message set successfully for " << user_principal_name << endl;
        }
    } catch (const exception& e) {
        cout << "Error setting auto-reply messages: " << e.what() << endl;
    }
}

int main() {
    read_csv_and_set_replies();
    return 0;
}
